File: SevenNightClassPCS/oop-cpp/animals/animals.cpp
#include <iostream>
using namespace std;

#include <vector>

class Animal {
public:
    virtual void sayHi() {
        cout << "Should not be here" << endl;
    }
};

class Cat: public Animal {
public:
    virtual void sayHi() {
        cout << "Meow!" << endl;
    }
};

class Dog: public Animal {
public:
    virtual void sayHi() {
        cout << "Woof!" << endl;
    }
};

void speak(Animal an) {
    an.sayHi();
}

void speakPoly(Animal * an) {
    an->sayHi();
}

int main(int argc, char **argv) {

    Dog * dog = new Dog();
    Cat * cat = new Cat();

    dog->sayHi();

    vector<Animal*> zoo;

    zoo.push_back(dog);
    zoo.push_back(cat);

    zoo[0]->sayHi();

    zoo.erase(zoo.begin() + 1);
    zoo.erase(zoo.begin());

    delete dog;
    delete cat;

}

File: Topics/02_Inheritance/02_05_JavaCollections/java_source/Animals.java
import java.util.ArrayList;
import java.util.List;

interface Animal {
	public void sayHi();
}

class Cat implements Animal {
	public void sayHi() {
		System.out.println("Meow!");		
	}	
}

class Dog implements Animal {
	public void sayHi() {
		System.out.println("Woof!");
	}	
}

public class Animals {
	
	public static void speak(Animal an) {
		an.sayHi();
	}
	
	public static void useList() {
		
		List names = new ArrayList();		
		
		names.add("Sam");
		names.add("Jan");
		
		String bob = "Bob";
		Object o = bob;
		
		names.add(o);
		
		String n = (String) names.get(1);
		
		System.out.println(names.get(1));
		System.out.println(names.get(2));		
		
		names.add(1,"Sue");
		
		System.out.println(names.get(0));
		System.out.println(names.get(1));
		
		System.out.println("There are "+names.size()+" items.");
		
	}
	
	public static void main(String [] args) {
		
		//useList();
		
		List<Integer> nums = new ArrayList<Integer>();
		
		nums.add(0);
		int a = nums.get(0);
		
		nums.add(null);
		int b = nums.get(1);
		
		
		List<Animal> zoo = new ArrayList<Animal>();
		
		Cat cat = new Cat();
		Dog dog = new Dog();
		
		speak(dog);
		
		zoo.add(cat);
		zoo.add(dog);
		
		Animal an = zoo.get(0);
		an.sayHi();
		
		an = zoo.get(1);
		an.sayHi();
		
	}

}
